src/arkode/arkode_mristep.c

```diff
@@ -1413,239 +1520,1440 @@ int mriStep_FullRHS(ARKodeMem ark_mem, sunrealtype t, N_Vector y, N_Vector f,
+  sunbooleantype calc_fslow;
+  sunbooleantype need_inner_dsm;
+  sunbooleantype do_embedding;
+  int nvec;
+
+  /* access the MRIStep mem structure */
+  retval = mriStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* initialize algebraic solver convergence flag to success;
+     error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* determine whether embedding stage is needed */
+  do_embedding = !ark_mem->fixedstep ||
+                 (ark_mem->AccumErrorType != ARK_ACCUMERROR_NONE);
+
+  /* if MRI adaptivity is enabled: reset fast accumulated error,
+     and send appropriate control parameter to the fast integrator */
+  adapt_type     = SUNAdaptController_GetType(ark_mem->hadapt_mem->hcontroller);
+  need_inner_dsm = SUNFALSE;
+  if (adapt_type == SUN_ADAPTCONTROLLER_MRI_H_TOL)
+  {
+    need_inner_dsm      = SUNTRUE;
+    step_mem->inner_dsm = ZERO;
+    retval = mriStepInnerStepper_ResetAccumulatedError(step_mem->stepper);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to reset the inner stepper error estimate");
+      return (ARK_INNERSTEP_FAIL);
+    }
+    retval = mriStepInnerStepper_SetRTol(step_mem->stepper,
+                                         step_mem->inner_rtol_factor *
+                                           ark_mem->reltol);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to set the inner stepper tolerance");
+      return (ARK_INNERSTEP_FAIL);
+    }
+  }
+
+  /* for adaptive computations, reset the inner integrator to the beginning of this step */
+  if (!ark_mem->fixedstep)
+  {
+    retval = mriStepInnerStepper_Reset(step_mem->stepper, ark_mem->tn,
+                                       ark_mem->yn);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to reset the inner stepper");
+      return (ARK_INNERSTEP_FAIL);
+    }
+  }
+
+  /* call nonlinear solver setup if it exists */
+  if (step_mem->NLS)
+  {
+    if ((step_mem->NLS)->ops->setup)
+    {
+      N_VConst(ZERO, ark_mem->tempv3); /* set guess to 0 */
+      retval = SUNNonlinSolSetup(step_mem->NLS, ark_mem->tempv3, ark_mem);
+      if (retval < 0) { return (ARK_NLS_SETUP_FAIL); }
+      if (retval > 0) { return (ARK_NLS_SETUP_RECVR); }
+    }
+  }
+
+  /* Evaluate the slow RHS functions if needed. NOTE: we decide between calling the
+     full RHS function (if ark_mem->fn is non-NULL) versus just updating the stored
+     values of Fse[0] and Fsi[0].  In either case, we use ARK_FULLRHS_START mode
+     because MRIGARK methods do not evaluate the RHS functions at the end of the
+     time step (so nothing can be leveraged). */
```

Comment:
For the "middle" integrators in a nested MRI method I think we need to call `UpdateF0` rather than `FullRHS` so the interpolation module only uses the RHS functions at the corresponding time scale for a higher-order predictor. I think we can check this by looking at the `step_mem->expforcing` and `step_mem->impforcing` flags i.e., updating the logic below to something like
```
  /* Is this a nested MRI configuration? */
  sunbooleantype nested_mri = step_mem->expforcing || step_mem->impforcing;

  if (ark_mem->fn == NULL || nested_mri)
  {
    retval = mriStep_UpdateF0(ark_mem, step_mem, ark_mem->tn, ark_mem->yn,
                              ARK_FULLRHS_START);
    if (retval) { return ARK_RHSFUNC_FAIL; }

    /* For a nested MRI configuration we might still need fn to create a predictor but
       it should be fn only for the current nesting level which is why we use UpdateF0
       in this case rather than FullRHS */
    if (ark_mem->fn != NULL && nested_mri && step_mem->implicit_rhs)
    {
      if (step_mem->implicit_rhs && step_mem->explicit_rhs)
      {
        N_VLinearSum(ONE, step_mem->Fsi[0], ONE, step_mem->Fse[0], ark_mem->fn);
      }
      else
      {
        N_VScale(ONE, step_mem->Fsi[0], ark_mem->fn);
      }
    }
  }
  else if (ark_mem->fn != NULL && !ark_mem->fn_is_current)
  {
    retval = mriStep_FullRHS(ark_mem, ark_mem->tn, ark_mem->yn, ark_mem->fn,
                             ARK_FULLRHS_START);
    if (retval) { return ARK_RHSFUNC_FAIL; }
  }
  ark_mem->fn_is_current = SUNTRUE;